### PR TITLE
layout: implement unpacking and validation

### LINKED ISSUE #75: oci-image-tool: validate and unpack image
The oci-image-tool should be able to validate and unpack an oci-image from 1. a directory and 2. a tar file.

Proposed spec: https://github.com/opencontainers/image-spec/issues/23#issuecomment-212636147

### PR DESCRIPTION
**Pull request by s-urbaniak**
Fixes #75

Signed-off-by: Sergiusz Urbaniak <sur@coreos.com>

### CODE REVIEW

**Comment by s-urbaniak** [comment]
Blocked by https://github.com/opencontainers/image-spec/pull/81: I need the config validator to thoroughly validate the image/image layout.

**Comment by philips** [comment]
@s-urbaniak #81 is merged, what is the next step?

**Comment by s-urbaniak** [comment]
@philips implementing image layout validation now, should be ready by EOD

**Comment by s-urbaniak** [comment]
@philips @vbatts @stevvooe PTAL; I implemented this using self-made image available here https://github.com/s-urbaniak/oci-image-test.

**Comment by philips** [comment]
@s-urbaniak Are you generating the OCI Runtime config.json? https://github.com/opencontainers/runtime-spec/blob/master/config.md We need to generate that from the serialization.config thingie.

**Comment by s-urbaniak** [comment]
@philips nope, as discussed OOB, this is just "pure" unpacking. The runtime config.json would have to be solved in a separate PR.

**Comment by stevvooe** [comment]
@s-urbaniak I'm looking at this today. Sorry for the delay.

File: image/unpack.go [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+			// success, continue below
+		default:
+			return errors.Wrapf(err, "error advancing tar stream")
+		}
+
+		path := filepath.Join(dest, filepath.Clean(hdr.Name))
+		info := hdr.FileInfo()
+
+		if strings.HasPrefix(info.Name(), `.wh.`) {
+			path = strings.Replace(path, `.wh.`, "", 1)
```
Comment (philips) [inline image/unpack.go]: any reason to use backtick instead of regular quotes?

**Comment by stevvooe** [comment]
@s-urbaniak This is looking very solid. Is there a corresponding specification for this format?

The only large change I'd suggest is that rather than `manifests`, this should be `refs`. These files should simply contain a descriptor pointing to the manifest blob. This will allow us to avoid the half-deserialized manifests.

File: image/unpack.go [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+			if info.IsDir() {
+				return nil
+			}
+
+			dd, err := filepath.Rel("blobs", filepath.Clean(path))
+			if err != nil || d != dd {
+				return nil // ignore
+			}
+
+			if err := extractLayer(dest, r); err != nil {
```
Comment (stevvooe) [inline image/unpack.go]: If we unpack directly to disk, these layers really need to be verified against the manifest's digest.

Comment (philips) [inline image/unpack.go]: +1 on that.

**Comment by philips** [comment]
@stevvooe what do you mean by "avoid the half-deserialized manifests"?

**Comment by philips** [comment]
@stevvooe are you thinking something like this for the refs schema:

```
$ cat refs/v1.0
{
  "media-types": {
    "application/vnd.oci.image.manifest.v1+json": {"size": "4096", "digest": "sha256:afff3928"
  }
}
```

**Comment by s-urbaniak** [comment]
@stevvooe @philips thanks a lot for the review! just fyi: I'm sitting between flights though so won't make to implement changes before tomorrow.

File: cmd/oci-image-tool/autodetect.go [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+	"net/http"
+	"os"
+
+	"github.com/opencontainers/image-spec/schema"
+	"github.com/pkg/errors"
+)
+
+// supported autodetection types
+const (
+	typeImageLayout  = "imageLayout"
```
Comment (stevvooe) [inline cmd/oci-image-tool/autodetect.go]: Why not use the mediatypes directly?

**Comment by stevvooe** [comment]
@philips I really referring to https://github.com/opencontainers/image-spec/pull/82/files#diff-1d4f0c730d731e154e77caf8c8ce36a4R80.

But, I was thinking about something like this:

```
$ cat refs/v1.0
{"size": 4096, "digest": "sha256:afff3928", "mediatype": "application/vnd.oci.image.manifest.v1+json"}
```

**Comment by philips** [comment]
@stevvooe I think we need an array of media types though as we need a way of emulating Accept with a filesystem. What if my ref is serving a v1 and v1.1 manifest? Let me file a new issue on this.

**Comment by wking** [comment]
On Thu, May 26, 2016 at 02:14:36PM -0700, Brandon Philips wrote:
> What if my ref is serving a v1 and v1.1 manifest?

Make the root reference point to a manifest.list [1]?

[1]: https://github.com/opencontainers/image-spec/blob/master/manifest.md#manifest-list

**Comment by philips** [comment]
@wking I think that is the approach we are going to take. Let me work up a proposal.

**Comment by philips** [comment]
@stevvooe here is a concrete format specification: https://github.com/opencontainers/image-spec/pull/94

File: image/unpack.go [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+	digests, err := diffIDs(w, manifest)
+	if err != nil {
+		return errors.Wrap(err, "error reading diff IDs")
+	}
+
+	return extractLayers(w, digests, dest)
+}
+
+func diffIDs(w walker, manifest string) ([]string, error) {
+	manifest = filepath.Join("manifests", manifest)
```
Comment (philips) [inline image/unpack.go]: can you change this code to follow the spec here: https://github.com/opencontainers/image-spec/pull/94

File: image/unpack.go (hunk 1 above)
Comment (s-urbaniak) [inline image/unpack.go]: no reason, just too fast typing ;-) fixed

File: cmd/oci-image-tool/autodetect.go (hunk 3 above)
Comment (s-urbaniak) [inline cmd/oci-image-tool/autodetect.go]: this gives me finer grained control, i.e. `typeImageLayout`, and `typeImage` don't have media types at all and must be treated differently.

File: image/unpack.go (hunk 2 above)
Comment (s-urbaniak) [inline image/unpack.go]: fixed

File: image/unpack.go (hunk 4 above)
Comment (s-urbaniak) [inline image/unpack.go]: fixed

**Comment by s-urbaniak** [comment]
@stevvooe @philips refactored this towards the proposed layout spec in https://github.com/opencontainers/image-spec/pull/94.

**Comment by philips** [comment]
Overall LGTM for a starting point. 

Anyone else give this an LGTM @opencontainers/image-spec-maintainers

File: cmd/oci-image-tool/autodetect.go (hunk 3 above)
Comment (stevvooe) [inline cmd/oci-image-tool/autodetect.go]: Of course!

**Comment by jonboulle** [comment]
LGTM

**Pull request merged**